Add Percolator section (#24)
tikv/deep-dive-tikv: src/distributed-transaction/percolator.md

# Percolator

> TiKV supports distributed transaction, which is inspired by Google's parer about *Percolator*. We will introduce it and how we make use of it in TiKV in this section.

Reviewer: paper?

## What is Percolator?

> *Percolator* is a system used for incremental processing on very large data set, built by Google. Support for distributed transactions is only a part of it. We will introduce Percolator briefly here. You can view the full paper [here](https://ai.google/research/pubs/pub36726#), and if you are very familiar to this parer, you can skip this section and go to read [Percolator in TiKV](#Percolator-in-TiKV)

Reviewer: paper

Percolator is built based on Google's BigTable. BigTable is a distributed storage system that supports single-row transactions, based on which Percolator implements distributed transactions in ACID snapshot-isolation semantics. A column `c` of Percolator is actually divided into following 5 internal columns of BigTable:

* `c:lock`
* `c:write`
* `c:data`
* `c:notify`
* `c:ack_O`

> `c:notify` and `c:ack_O` are used for Percolator's incremental processing. They are not used by TiKV, so let's ignore these two columns and consider only the first three columns.

Reviewer: Since in this section, we are talking about the standard Percolator. Is it appropriate to ignore these two columns?

Author: Emmmm.. Maybe we should keep them and explain them in a short paragraph. I'll write it later.

> Percolator also relies on a service named *timestamp oracle*. The timestamp oracle can produces timestamps in strictly increasing order. All read and write operations needs to apply for timestamps from the timestamp oracle, and a timestamp came from timestamp oracle will be regarded as the logical time when the read/write operation happens.

Reviewer: ```suggestion
Percolator also relies on a service named *timestamp oracle*. The timestamp oracle can produce timestamps in strictly increasing order. All read and write operations need to apply for timestamps from the timestamp oracle, and a timestamp come from timestamp oracle will be regarded as the logical time when the read/write operation happens.
```

Percolator is multi-version, and a data item's version is represented by the timestamp that the transaction was committed.

For example, in such a state:

<table>
  <tr>
    <th>key</th><th>v:data</th><th>v:lock</th><th>v:write</th>
  </tr>
  <tr>
    <td rowspan="3">k1</td>
    <td>14: "value2"</td><td>14: primary</td><td></td>
  </tr>
  <tr>
    <td></td><td></td><td>12: data@10</td>
  </tr>
  <tr>
    <td>10: "value1"</td><td></td><td></td>
  </tr>
</table>

It means that for the key `k1`, a value `"value1"` was committed at timestamp `12`. Then there is an uncommitted version whose value is `"value2"`, and it's uncommitted because there's a lock. You will understand why it looks like this after understanding how transactions works.

### Writing

Percolator's transactions are committed by a 2PC algorithm. Its two phases are called `Prewrite` and `Commit`. 

In `Prewrite` phase:
1. We get a timestamp from the timestamp oracle, and we call the timestamp the transaction's `start_ts`.

> 2. For each rows involved in the transaction, put a lock to `lock` column. At the same time, put the value to the `data` column with the timestamp `start_ts`. One of these locks will be chose as the *primary* lock, and others are *secondary* locks. Each lock saves the transaction's `start_ts`. The secondary locks, in addition, saves the primary lock's location.

Reviewer: ```suggestion
2. For each row involved in the transaction, put a lock to `lock` column. At the same time, put the value to the `data` column with the timestamp `start_ts`. One of these locks will be chosen as the *primary* lock, and others are *secondary* locks. Each lock saves the transaction's `start_ts`. The secondary locks, in addition, save the primary lock's location.
```

    * If we found that there's already a lock or newer version than `start_ts`, the current transaction will be rolled back.

And then, in `Commit` phase:
1. Get another timestamp, namely `commit_ts`.

> 2. Remove the primary lock, and at the same time write a record to `write` column with the timestamp `commit_ts`, and it's value records that the transaction's `start_ts`.

Reviewer: ```suggestion
2. Remove the primary lock, and at the same time write a record to `write` column with the timestamp `commit_ts`, and its value records the transaction's `start_ts`.
```

# Percolator

> TiKV supports distributed transaction, which is inspired by Google's paper about *Percolator*. We will introduce it and how we make use of it in TiKV in this section.

Reviewer: ```suggestion
TiKV supports the distributed transaction, which is inspired by Google's paper about *Percolator*. We will introduce it and how we make use of it in TiKV in this section.
```

## What is Percolator?

*Percolator* is a system used for incremental processing on very large data set, built by Google. Support for distributed transactions is only a part of it. We will introduce Percolator briefly here. You can view the full paper [here](https://ai.google/research/pubs/pub36726#), and if you are very familiar to this paper, you can skip this section and go to read [Percolator in TiKV](#Percolator-in-TiKV)

Percolator is built based on Google's BigTable. BigTable is a distributed storage system that supports single-row transactions, based on which Percolator implements distributed transactions in ACID snapshot-isolation semantics. A column `c` of Percolator is actually divided into following internal columns of BigTable:

* `c:lock`
* `c:write`
* `c:data`
* `c:notify`
* `c:ack_O`

> Percolator also relies on a service named *timestamp oracle*. The timestamp oracle can produces timestamps in strictly increasing order. All read and write operations need to apply for timestamps from the timestamp oracle, and a timestamp came from timestamp oracle will be regarded as the logical time when the read/write operation happens.

Reviewer: produce

Percolator is multi-version, and a data item's version is represented by the timestamp that the transaction was committed.

For example, in such a state:

<table>
  <tr>
    <th>key</th><th>v:data</th><th>v:lock</th><th>v:write</th>
  </tr>
  <tr>
    <td rowspan="3">k1</td>
    <td>14: "value2"</td><td>14: primary</td><td></td>
  </tr>
  <tr>
    <td></td><td></td><td>12: data@10</td>
  </tr>
  <tr>
    <td>10: "value1"</td><td></td><td></td>
  </tr>
</table>

> It means that for the key `k1`, a value `"value1"` was committed at timestamp `12`. Then there is an uncommitted version whose value is `"value2"`, and it's uncommitted because there's a lock. You will understand why it looks like this after understanding how transactions works.

Reviewer: ```suggestion
It means that for the key `k1`, a value `"value1"` was committed at timestamp `12`. Then there is an uncommitted version whose value is `"value2"`, and it's uncommitted because there's a lock. You will understand why it looks like this after understanding how transactions work.
```

The data of a certain row and a certain Percolator column is named a *cell*. The above example shows different versions of data of a single cell.

> The remaining columns, `c:notify` and `c:ack_O`, are used for Percolator's incremental processing. After a modification, we set `c:notify` column to mark a cell to be dirty. There can be some *observers* which can does some user-specified operations when they finds data of their observed columns has changed. To find whether data is changed, they continuously scanning the `notify` columns to find dirty cells. `c:ack_O` is the "acknowledgement" column of the observer `O`, which is used to prevent a row from incorrectly being notified twice. It saves the timestamp of the observer's last execution.

Reviewer: ```suggestion
The remaining columns, `c:notify` and `c:ack_O`, are used for Percolator's incremental processing. After a modification, we set `c:notify` column to mark a cell to be dirty. There can be some *observers* which can do some user-specified operations when they find data of their observed columns have changed. To find whether data is changed, they continuously scanning the `notify` columns to find dirty cells. `c:ack_O` is the "acknowledgment" column of the observer `O`, which is used to prevent a row from incorrectly being notified twice. It saves the timestamp of the observer's last execution.
```

### Writing

Percolator's transactions are committed by a 2PC algorithm. Its two phases are called `Prewrite` and `Commit`.

In `Prewrite` phase:
1. We get a timestamp from the timestamp oracle, and we call the timestamp the transaction's `start_ts`.

> 2. For each row involved in the transaction, put a lock to `lock` column. At the same time, put the value to the `data` column with the timestamp `start_ts`. One of these locks will be chosen as the *primary* lock, and others are *secondary* locks. Each lock save the transaction's `start_ts`. The secondary locks, in addition, saves the primary lock's location.

Reviewer 2: ```suggestion
2. For each row involved in the transaction, put a lock to `lock` column. At the same time, put the value to the `data` column with the timestamp `start_ts`. One of these locks will be chosen as the *primary* lock, and others are *secondary* locks. Each lock saves the transaction's `start_ts`. The secondary locks, in addition, save the primary lock's location.
```

English is weird. 😵

    * If we found that there's already a lock or newer version than `start_ts`, the current transaction will be rolled back.

And then, in `Commit` phase:
1. Get another timestamp, namely `commit_ts`.
2. Remove the primary lock, and at the same time write a record to `write` column with the timestamp `commit_ts`, and its value records the transaction's `start_ts`.
    * If the lock is missing, the commit should fail.
3. For all secondaries, do the same thing.

Once the step 2 (committing the primary) is done, we say the whole transaction is done. It doesn't matter if the process of committing the secondaries unexpectedly exited. When we meet a lock of other transactions, we can see whether its primary has been committed to determine whether the whole transaction has been committed.

Let's see the example in the paper of Percolator. Assume we are writing two rows in a single transaction. At first, the data looks like this:

> ![Figure 1](percolator-figure-1.png)

Reviewer 3: em, can we use markdown?

Reviewer 3: Like

| key  | bal:data      | Bal:lock   | Bal:write        |
| ---- | ------------- | ---------- | ---------------- |
| Bob  | 6<br />5: $10 | 6:<br />5: | 6:data@5<br />5: |
| Joe  | 6:<br />5:$2  | 6:<br />5: | 6:data@5<br />5: |

The code is

```
| key  | bal:data      | Bal:lock   | Bal:write        |
| ---- | ------------- | ---------- | ---------------- |
| Bob  | 6<br />5: $10 | 6:<br />5: | 6:data@5<br />5: |
| Joe  | 6:<br />5:$2  | 6:<br />5: | 6:data@5<br />5: |
```

Author: Did you mean drawing the table using Markdown? Using Markdown is difficult to draw row spanning in tables, which is useful to represent multiple versions of one cell. I tried to draw it directly in HTML, but then I found the code kind of ugly...

Author: Oh, I've tried this approach too, but I gave up because I'm afraid it's ugly to let there be so many `<br/>`s... I'll adopt this way again since you think it's ok

Reviewer 3: IMO, it may be easier if you want to update the table later. 

You use different two styles in one documents, above is using pure HTML table, below is using the picture, seem we should unify them :-)

Author: ok.

Author: updated..PTAL

> This table shows Bob and Joe's balance. Now Bob want to transfer his $7 to Joe's account. First, we do `Prewrite`:

Reviewer: ```suggestion
This table shows Bob and Joe's balance. Now Bob wants to transfer his $7 to Joe's account. First, we do `Prewrite`:
```

1. Get the `start_ts` of the transaction. We got `7` in this example.

> 2. For each row involved in this transaction, put a lock in the `lock` column, and put the data to the `data` column. One of the locks will be chose as the primary lock.

Reviewer: ```suggestion
2. For each row involved in this transaction, put a lock in the `lock` column, and put the data to the `data` column. One of the locks will be chosen as the primary lock.
```

After `Prewrite`, our data looks like this:

![Figure 2](percolator-figure-2.png)

Then `Commit`:

1. Get the `commit_ts`. We got `8`.
2. Commit the primary: Remove the primary lock and put the commit record to `write` column. We get this:
    ![Figure 3](percolator-figure-3.png)
3. Commit all secondaries. Finally, we get this:
    ![Figure 4](percolator-figure-4.png)

### Reading

To read from Percolator, a timestamp is required too. The procedure to perform read operation is as following:

1. Get a timestamp, `ts`.

> 2. Check if the row we are going to read is locked with timestamp in the range `[0, ts]`;

Reviewer: ```suggestion
2. Check if the row we are going to read is locked with a timestamp in the range `[0, ts]`;
```

>     * If so, it means the row is locked by a earlier-started transaction. Then we are not sure whether that transaction will be committed before `ts` or after `ts`. Other wise, we can continue reading.

Reviewer: ```suggestion
    * If so, it means the row is locked by an earlier-started transaction. Then we are not sure whether that transaction will be committed before `ts` or after `ts`. Otherwise, we can continue reading.
```

3. Get the latest record in the row's `write` column whose `commit_ts` is in range `[0, ts]`. The record contains the `start_ts` of the transaction where it was committed.
4. Get the row's value in `data` column, whose timestamp is exactly `start_ts`. Then the value is what we want.

> For example, consider a table as showed in the image again:

Reviewer: ```suggestion
For example, consider a table as shown in the image again:
```

![Figure 5](percolator-figure-3.png)

Let's read Bob's balance.
1. Get a timestamp. Assume we got `9`.
2. Check lock. The row of Bob is not locked, so we continue.
3. Get the latest record in `write` column that committed before `9`. We got the record with `commit_ts` equals to `8`, and it says the `start_ts` is `7`, which means, it's corresponding data is at timestamp `7` in `data` column.
4. Get the value in `data` column whose timestamp is exactly `7`. We got `$3`.

> This algorithm provides us both lock-free reading and the ability to perform historical read. In the above example, if we specify that we want to read at time point `7`, then we will se the write record at timestamp `6`, giving us the result `$10` at timestamp `5`.

Reviewer: se?

### Handling Conflicts

Conflicts are identified by checking the `lock` column. A row can have many versions of data, but it can have at most one lock at any time.

> When we are performing a write operation, we tries to lock every affected rows in the `Prewrite` phase. If we failed to lock some of these rows, the whole transaction will be rolled back. As an optimistic lock algorithm, sometimes it causes a shortage of Percolator: there may be performance regression in a scenario that conflicts occur frequently.

Reviewer: ```suggestion
When we are performing a write operation, we try to lock every affected row in the `Prewrite` phase. If we failed to lock some of these rows, the whole transaction will be rolled back. As an optimistic lock algorithm, sometimes it causes a shortage of Percolator: there may be performance regression in a scenario that conflicts occur frequently.
```

To roll back a row, just simply remove its lock and its corresponding value in `data` column.

### Tolerating crashes

> Percolator has the ability to surfer crushing without breaking correctness of data.

Reviewer 2: ```suggestion
Percolator has the ability to survive crashes without breaking correctness of data.
```

> Percolator has the ability to survive crashes without breaking correctness of data.

Reviewer: ```suggestion
Percolator has the ability to survive crashes without breaking the correctness of data.
```

> First, let's see what will happen after a crash. A crash may happen during `Prewrite`, during `Commit` or between these two. But any way, we can simply divide these conditions into two types: before committing the primary, or after committing the primary.

Reviewer: ```suggestion
First, let's see what will happen after a crash. A crash may happen during `Prewrite`, during `Commit` or between these two. But anyway, we can simply divide these conditions into two types: before committing the primary, or after committing the primary.
```

> So, when a transaction `T1` (either reading or writing) finds that a row `R1` has a lock which belongs to an earlier transaction `T0`, it in fact doesn't simply rollback itself immediately. However, it checks the state of the lock's primary.

Reviewer: ```suggestion
So, when a transaction `T1` (either reading or writing) finds that a row `R1` has a lock which belongs to an earlier transaction `T0`, it, in fact, doesn't simply rollback itself immediately. However, it checks the state of the lock's primary.
```

* If the primary lock disappeared but there's a record `data @ T0.start_ts` in `write` column, it means that `T0` has been successfully committed. Then the row `R1`'s stale lock can also be committed. Usually we call this `rolling forward`. After that, the new transaction `T1` can continue.
* If the primary lock disappeared with nothing left, it means the transaction has been rolled back. Then the row `R1`'s stale lock should also be rolled back. Then `T1` can continue.
* If the primary lock exists but it's too old (we can set a TTL to the lock), we can believe that the transaction crashed before committing or rolling back. We can rollback it and continue `T1`.

> * Otherwise, we consider transaction `T0` is still running. Then `T1` can rollback it self, or another choice, try to wait for a while to see whether `T0` will be committed before `T1.start_ts`.

Reviewer: ```suggestion
* Otherwise, we consider transaction `T0` is still running. Then `T1` can rollback itself, or another choice, try to wait for a while to see whether `T0` will be committed before `T1.start_ts`.
```

> * Other wise, we consider transaction `T0` is still running. Then `T1` can rollback it self, or another choice, try to wait for a while to see whether `T0` will be committed before `T1.start_ts`.

Reviewer 2: ```suggestion
* Otherwise, we consider transaction `T0` is still running. Then `T1` can rollback itself, or another choice, try to wait for a while to see whether `T0` will be committed before `T1.start_ts`.
```

## Percolator in TiKV

TiKV is a key-value storage, and each key-value pair can be regarded as a row in Percolator.

TiKV internally uses RocksDB, a key-value storage engine library, to persist data to disk. RocksDB's atomic write batch and TiKV's transaction scheduler make it atomic to read and write a single user key, which is a requirement of Percolator.

> RocksDB provides a features named *Column Family* (we will abbreviate it as *CF*). An instance of RocksDB may have multiple CFs, and each CF is a separated key namespace and has its own LSM-Tree. However different CFs in the same RocksDB instance uses a common WAL, providing the ability to write to different CFs atomically.

Reviewer: ```suggestion
RocksDB provides a feature named *Column Family* (we will abbreviate it as *CF*). An instance of RocksDB may have multiple CFs, and each CF is a separated key namespace and has its own LSM-Tree. However different CFs in the same RocksDB instance uses a common WAL, providing the ability to write to different CFs atomically.
```

We divide a RocksDB to three CFs: `CF_DEFAULT`, `CF_LOCK` and `CF_WRITE`, which corresponds to Percolator's `data` column, `lock` column and `write` column respectively. There's an extra CF named `CF_RAFT` used to save some metadata of Raft, but that's out of our topic now. The `notify` and `ack_O` columns are not present in TiKV, because for now TiKV doesn't need the ability of incremental processing.

> Then, we need to represent different versions of a key. We can simply compound a key and a timestamp as an internal key, and use the internal key in RocksDB. But since a key can have at most one lock at a time, so we don't need to add timestamp to the key in `CF_LOCK`. Then the contents of each CFs are:

Reviewer: ```suggestion
Then, we need to represent different versions of a key. We can simply compound a key and a timestamp as an internal key, and use the internal key in RocksDB. But since a key can have at most one lock at a time, so we don't need to add a timestamp to the key in `CF_LOCK`. Then the contents of each CFs are:
```

* `CF_DEFAULT`: `(key, start_ts)` -> `value`
* `CF_LOCK`: `key` -> `lock_info`
* `CF_WRITE`: `(key, commit_ts)` -> `write_info`

Our technical to compose user keys and timestamps together is:

1. Encode the user key to [memcomparable](https://github.com/facebook/mysql-5.6/wiki/MyRocks-record-format#memcomparable-format)
2. Bitwise invert the timestamp (which is an unsigned int64) and encode it to big-endian bytes
3. Append the encoded timestamp to the encoded key.

For example, the key `"key1"` and the timestamp `3` will be encoded as `"key1\x00\x00\x00\x00\xfb\xff\xff\xff\xff\xff\xff\xff\xfe"`, where the first 9 bytes is the memcomparable-encoded key and the remaining 8 bytes is the inverted timestamp in big-endian. In this way, our different versions of the same key are always adjacent in RocksDB, and for each key, newer versions are always in front of older versions.

> There are some differences between TiKV and the Percolator's paper. In tikv, records in `CF_WRITE` has four different types: `Put`, `Delete`, `Rollback` and `Lock`. Only `Put` records need a corresponding value in `CF_DEFAULT`. When rolling back transactions, we doesn't simply remove the lock but writes a `Rollback` record in `CF_WRITE`. The `Lock` type of write records is different from Percolator's lock, but is produced by queries like `SELECT ... FOR UPDATE` in TiDB. For keys affected by this query, although they are only read, it's part of a write operation. To guarantee to be in snapshot-isolation, we make it acts like a write operation (though it doesn't write anything) to ensure the keys are locked and won't change before committing the transaction.

Reviewer: ```suggestion
There are some differences between TiKV and the Percolator's paper. In TiKV, records in `CF_WRITE` has four different types: `Put`, `Delete`, `Rollback` and `Lock`. Only `Put` records need a corresponding value in `CF_DEFAULT`. When rolling back transactions, we don't simply remove the lock but writes a `Rollback` record in `CF_WRITE`. The `Lock` type of write records is different from Percolator's lock, but is produced by queries like `SELECT ... FOR UPDATE` in TiDB. For keys affected by this query, although they are only read, it's part of a write operation. To guarantee to be in snapshot-isolation, we make it acts like a write operation (though it doesn't write anything) to ensure the keys are locked and won't change before committing the transaction.
```

# Percolator

> TiKV supports the distributed transaction, which is inspired by Google's paper about *Percolator*. We will introduce it and how we make use of it in TiKV in this section.

Reviewer 4: ```suggestion
TiKV supports distributed transactions, which is inspired by Google's *Percolator*. In this section, we will briefly introduce Percolator and how we make use of it in TiKV.
```

## What is Percolator?

> *Percolator* is a system used for incremental processing on very large data set, built by Google. Support for distributed transactions is only a part of it. We will introduce Percolator briefly here. You can view the full paper [here](https://ai.google/research/pubs/pub36726#), and if you are very familiar to this paper, you can skip this section and go to read [Percolator in TiKV](#Percolator-in-TiKV)

Reviewer 4: ```suggestion
*Percolator* is a system built by Google for incremental processing on a very large data set. Since this is just a brief introduction, you can view the full paper [here](https://ai.google/research/pubs/pub36726#) for more details. If you are already very familiar with it, you can skip this section and go directly to [Percolator in TiKV](#Percolator-in-TiKV)
```

> Percolator is built based on Google's BigTable. BigTable is a distributed storage system that supports single-row transactions, based on which Percolator implements distributed transactions in ACID snapshot-isolation semantics. A column `c` of Percolator is actually divided into following internal columns of BigTable:

Reviewer 4: ```suggestion
Percolator is built based on Google's BigTable, a distributed storage system that supports single-row transactions. With similar logic, Percolator implements distributed transactions in ACID snapshot-isolation semantics. A column `c` of Percolator is actually divided into the following internal columns of BigTable:
```

Author: Not "with similar logic". BigTable doesn't support ACID distributed transaction, and Percolator implemented it based on BigTable. How do you think if I change this paragraph to this?
> Percolator is built based on Google's BigTable, a distributed storage system that supports single-row transactions. Percolator implements distributed transactions in ACID snapshot-isolation semantics, which is not supported by BigTable. ...

Reviewer 4: Yeah it's accurate. At first I just wanted to avoid using "based on" twice in two sentences in a row. Go ahead and change it.

* `c:lock`
* `c:write`
* `c:data`
* `c:notify`
* `c:ack_O`

> Percolator also relies on a service named *timestamp oracle*. The timestamp oracle can produce timestamps in strictly increasing order. All read and write operations need to apply for timestamps from the timestamp oracle, and a timestamp came from timestamp oracle will be regarded as the logical time when the read/write operation happens.

Reviewer 4: ```suggestion
Percolator also relies on a service named *timestamp oracle*. The timestamp oracle can produce timestamps in a strictly increasing order. All read and write operations need to apply for timestamps from the timestamp oracle, and a timestamp coming from the timestamp oracle will be regarded as the logical time when the read/write operation happens.
```

> Percolator is multi-version, and a data item's version is represented by the timestamp that the transaction was committed.

Reviewer 4: ```suggestion
Percolator supports multi-version, and a data item's version is represented by the timestamp when the transaction was committed.
```

Author: I wanted to use "multi-version" as an adjective. If that's weird, how is it if I say "Percolator is a multi-version storage"?

For example, in such a state:

| key | v:data | v:lock | v"write |
|-----|--------|--------|---------|
|k1   |14:"value2"<br/>12:<br/>10:"value1"|14:primary<br/>12:<br/>10:|14:<br/>12:data@10<br/>10:|

> It means that for the key `k1`, a value `"value1"` was committed at timestamp `12`. Then there is an uncommitted version whose value is `"value2"`, and it's uncommitted because there's a lock. You will understand why it looks like this after understanding how transactions work.

Reviewer 4: ```suggestion
The state in the table means that for key `k1`, value `"value1"` was committed at timestamp `12`. Then there is an uncommitted version whose value is `"value2"`, and it's uncommitted because there's a lock. You will understand why it is like this after understanding how transactions work.
```

> The data of a certain row and a certain Percolator column is named a *cell*. The above example shows different versions of data of a single cell.

Reviewer 4: ```suggestion
The data of a specific row and a specific Percolator column is named a *cell*. The above example shows different versions of data for a single cell.
```

> The remaining columns, `c:notify` and `c:ack_O`, are used for Percolator's incremental processing. After a modification, we set `c:notify` column to mark a cell to be dirty. There can be some *observers* which can do some user-specified operations when they find data of their observed columns have changed. To find whether data is changed, they continuously scanning the `notify` columns to find dirty cells. `c:ack_O` is the "acknowledgment" column of the observer `O`, which is used to prevent a row from incorrectly being notified twice. It saves the timestamp of the observer's last execution.

Reviewer 4: Are we supposed to introduce Percolator in this section? But "After a modification, we set `c:notify` column to..." looks like we are talking about our implementation in TiKV at the same time.

Author: Sorry. Is it ok to remove the "we", and use third person instead, such as "After a modification, `c:notify` column is used to mark the modified cell to be dirty"?

Reviewer 4: Makes sense.

> The remaining columns, `c:notify` and `c:ack_O`, are used for Percolator's incremental processing. After a modification, we set `c:notify` column to mark a cell to be dirty. There can be some *observers* which can do some user-specified operations when they find data of their observed columns have changed. To find whether data is changed, they continuously scanning the `notify` columns to find dirty cells. `c:ack_O` is the "acknowledgment" column of the observer `O`, which is used to prevent a row from incorrectly being notified twice. It saves the timestamp of the observer's last execution.

Reviewer 4: ```suggestion
The remaining columns, `c:notify` and `c:ack_O`, are used for Percolator's incremental processing. After a modification, we set `c:notify` column to mark a cell to be dirty. Existing *observers*  will perform user-specified operations when they find data of their observed columns has changed. To find whether data is changed, they continuously scan the `notify` columns to find dirty cells. `c:ack_O` is the "acknowledgment" column of observer `O`, which is used to prevent a row from being incorrectly notified twice. It saves the timestamp of the observer's last execution.
```

Author: The term "observer" occurs for the first time here (that's why I make it italic). I would like to change this sentence to "Users can add some *observers* to Percolator which can do user-specified operations....". Do you think that's ok?

### Writing

> Percolator's transactions are committed by a 2PC algorithm. Its two phases are called `Prewrite` and `Commit`.

Reviewer 4: ```suggestion
Percolator's transactions are committed by a 2-phase commit (2PC) algorithm. Its two phases are `Prewrite` and `Commit`.
```

In `Prewrite` phase:

> 1. We get a timestamp from the timestamp oracle, and we call the timestamp the transaction's `start_ts`.

Reviewer 4: Who is "we"?

Author: Sorry. Is it ok to say "Get a timestamp..." directly?

Reviewer 4: Yes.

> 2. For each row involved in the transaction, put a lock to `lock` column. At the same time, put the value to the `data` column with the timestamp `start_ts`. One of these locks will be chosen as the *primary* lock, and others are *secondary* locks. Each lock saves the transaction's `start_ts`. The secondary locks, in addition, save the primary lock's location.

Reviewer 4: ```suggestion
2. For each row involved in the transaction, put a lock in the `lock` column. At the same time, assign the value to the `data` column with the timestamp `start_ts`. One of these locks will be chosen as the *primary* lock, while others are *secondary* locks. Each lock contains the transaction's `start_ts`. Each secondary lock, in addition, contains the location of the primary lock.
```

Author: Sorry but I don't think "assign" is proper here. I prefer keeping the word "put".

>     * If we found that there's already a lock or newer version than `start_ts`, the current transaction will be rolled back.

Reviewer 4: ```suggestion
    * If there's already a lock or newer version than `start_ts`, the current transaction will be rolled back.
```

> And then, in `Commit` phase:

Reviewer 4: ```suggestion
And then, in the`Commit` phase:
```

1. Get another timestamp, namely `commit_ts`.

> 2. Remove the primary lock, and at the same time write a record to `write` column with the timestamp `commit_ts`, and its value records the transaction's `start_ts`.

Reviewer 4: ```suggestion
2. Remove the primary lock, and at the same time write a record to the `write` column with timestamp `commit_ts`, whose value records the transaction's `start_ts`.
```

>     * If the lock is missing, the commit should fail.

Reviewer 4: which lock is missing? The primary lock to remove? Also if there is only one point, we may consider moving it to the previous line instead of having one bullet for it.

> 3. For all secondaries, do the same thing.

Reviewer 4: ```suggestion
3. For all secondary locks, repeat the process above.
```

> Once the step 2 (committing the primary) is done, we say the whole transaction is done. It doesn't matter if the process of committing the secondaries unexpectedly exited. When we meet a lock of other transactions, we can see whether its primary has been committed to determine whether the whole transaction has been committed.

Reviewer 4: Consider removing "When we meet a lock of other transactions, we can see whether its primary has been committed to determine whether the whole transaction has been committed."?  I think it's a repetition to what has been said right before it.
```

> Let's see the example in the paper of Percolator. Assume we are writing two rows in a single transaction. At first, the data looks like this:

Reviewer 4: ```suggestion
Let's see an example from the paper of Percolator. Assume we are writing two rows in a single transaction. At first, the data looks like this:
```

Author: The paper provides only one example to explain the transaction. So I think it's ok to say "the" instead of "an"?

| key | bal:data    | bal:lock  | bal:write       |
|-----|-------------|-----------|-----------------|

> | Bob | 6:<br>5:$10 | 6:<br/>5: | 6:data@5<br/>5: |

Reviewer 3: ```
6:<br/>5:$10
```

| Joe | 6:<br/>5:$2 | 6:<br/>5: | 6:data@5<br/>5: |

> This table shows Bob and Joe's balance. Now Bob wants to transfer his $7 to Joe's account. First, we do `Prewrite`:

Reviewer 4: ```suggestion
This table shows Bob and Joe's balance. Now Bob wants to transfer his $7 to Joe's account. The first step is `Prewrite`:
```

Reviewer 4: I think we should pay attention to using a consistent and right person in our writing. Or else it would be somewhat confusing.

Author: Sorry for that 😢

> 1. Get the `start_ts` of the transaction. We got `7` in this example.

Reviewer 4: ```suggestion
1. Get the `start_ts` of the transaction. In our example, it's `7`.

> 2. For each row involved in this transaction, put a lock in the `lock` column, and put the data to the `data` column. One of the locks will be chosen as the primary lock.

Reviewer 4: ```suggestion
2. For each row involved in this transaction, put a lock in the `lock` column, and assign the data to the `data` column. One of the locks will be chosen as the primary lock.
```

Author: Could you please tell me why "assign" is better?

Reviewer 4: use write for similar instances

After `Prewrite`, our data looks like this:

| key | bal:data | bal:lock | bal:write |
|-----|----------|----------|-----------|
| Bob | 7:$3<br/>6:<br>5:$10 | 7:primary<br/>6:<br/>5:         | 7:<br/>6:data@5<br/>5: |
| Joe | 7:$9<br/>6:<br/>5:$2 | 7:primary@Bob.bal<br/>6:<br/>5: | 7:<br/>6:data@5<br/>5: |

Then `Commit`:

> 1. Get the `commit_ts`. We got `8`.

Reviewer 4: ```suggestion
1. Get the `commit_ts`, in our case, `8`.
```

> 2. Commit the primary: Remove the primary lock and put the commit record to `write` column. We get this:

Reviewer 4: ```suggestion
2. Commit the primary: Remove the primary lock and write the commit record to the `write` column. 
```

    | key | bal:data | bal:lock | bal:write |
    |-----|----------|----------|-----------|
    | Bob | 8:<br/>7:$3<br/>6:<br>5:$10 | 8:<br/>7:<br/>6:<br/>5: | 8:data@7<br/>7:<br/>6:data@5<br/>5: |
    | Joe | 7:$9<br/>6:<br/>5:$2 | 7:primary@Bob.bal<br/>6:<br/>5: | 7:<br/>6:data@5<br/>5: |

> 3. Commit all secondaries. Finally, we get this:

Reviewer 4: ```suggestion
3. Commit all secondary locks to complete the writing process.
```

    | key | bal:data | bal:lock | bal:write |
    |-----|----------|----------|-----------|
    | Bob | 8:<br/>7:$3<br/>6:<br>5:$10 | 8:<br/>7:<br/>6:<br/>5: | 8:data@7<br/>7:<br/>6:data@5<br/>5: |
    | Joe | 8:<br/>7:$9<br/>6:<br/>5:$2 | 8:<br/>7:<br/>6:<br/>5: | 8:data@7<br/>7:<br/>6:data@5<br/>5: |

### Reading

> To read from Percolator, a timestamp is required too. The procedure to perform read operation is as following:

Reviewer 4: ```suggestion
Reading from Percolator also requires a timestamp. The procedure to perform a read operation is as follows:
```

So, when a transaction `T1` (either reading or writing) finds that a row `R1` has a lock which belongs to an earlier transaction `T0`, it, in fact, doesn't simply rollback itself immediately. However, it checks the state of the lock's primary.

* If the primary lock disappeared but there's a record `data @ T0.start_ts` in `write` column, it means that `T0` has been successfully committed. Then the row `R1`'s stale lock can also be committed. Usually we call this `rolling forward`. After that, the new transaction `T1` can continue.
* If the primary lock disappeared with nothing left, it means the transaction has been rolled back. Then the row `R1`'s stale lock should also be rolled back. Then `T1` can continue.
* If the primary lock exists but it's too old (we can set a TTL to the lock), we can believe that the transaction crashed before committing or rolling back. We can rollback it and continue `T1`.
* Otherwise, we consider transaction `T0` is still running. Then `T1` can rollback itself, or another choice, try to wait for a while to see whether `T0` will be committed before `T1.start_ts`.

## Percolator in TiKV

> TiKV is a key-value storage, and each key-value pair can be regarded as a row in Percolator.

Reviewer 4: ```suggestion
TiKV is a distributed transactional key-value storage engine. Each key-value pair can be regarded as a row in Percolator.
```

> TiKV internally uses RocksDB, a key-value storage engine library, to persist data to disk. RocksDB's atomic write batch and TiKV's transaction scheduler make it atomic to read and write a single user key, which is a requirement of Percolator.

Reviewer 4: ```suggestion
TiKV internally uses RocksDB, a stand-alone key-value storage engine library, to persist data to local disk. RocksDB's atomic write batch and TiKV's transaction scheduler jointly make it atomic to read and write a single user key, which is a requirement of Percolator.
```

Author: RocksDB is a library but not a stand-alone process.
"atomic" is a term to say an operation is "non-dividable".

Reviewer 4: Got it. Then the original sentence is fine.

> TiKV internally uses RocksDB, a key-value storage engine library, to persist data to disk. RocksDB's atomic write batch and TiKV's transaction scheduler make it atomic to read and write a single user key, which is a requirement of Percolator.

Reviewer 4: ```suggestion
TiKV internally uses RocksDB, a stand-alone key-value storage engine library, to persist data to local disk. RocksDB's atomic write batch and TiKV's transaction scheduler jointly make it atomic to read and write a single user key, which is a requirement of Percolator.
```

> RocksDB provides a feature named *Column Family* (we will abbreviate it as *CF*). An instance of RocksDB may have multiple CFs, and each CF is a separated key namespace and has its own LSM-Tree. However different CFs in the same RocksDB instance uses a common WAL, providing the ability to write to different CFs atomically.

Reviewer 4: ```suggestion
RocksDB provides a feature named *Column Family* (hereafter referred to as *CF*). An instance of RocksDB may have multiple CFs, and each CF is a separated key namespace and has its own LSM-Tree. However different CFs in the same RocksDB instance uses a common WAL, providing the ability to write to different CFs atomically.
```

> We divide a RocksDB to three CFs: `CF_DEFAULT`, `CF_LOCK` and `CF_WRITE`, which corresponds to Percolator's `data` column, `lock` column and `write` column respectively. There's an extra CF named `CF_RAFT` used to save some metadata of Raft, but that's out of our topic now. The `notify` and `ack_O` columns are not present in TiKV, because for now TiKV doesn't need the ability of incremental processing.

Reviewer 4: ```suggestion
We divide a RocksDB instance to three CFs: `CF_DEFAULT`, `CF_LOCK` and `CF_WRITE`, which corresponds to Percolator's `data` column, `lock` column and `write` column respectively. There's an extra CF named `CF_RAFT` which is used to save some metadata of Raft, but that's beside our topic. The `notify` and `ack_O` columns are not present in TiKV, because for now TiKV doesn't need the ability of incremental processing.
```

> Then, we need to represent different versions of a key. We can simply compound a key and a timestamp as an internal key, and use the internal key in RocksDB. But since a key can have at most one lock at a time, so we don't need to add a timestamp to the key in `CF_LOCK`. Then the contents of each CFs are:

Reviewer 4: ```suggestion
Then, we need to represent different versions of a key. We can simply compound a key and a timestamp as an internal key, which can be used in RocksDB. However, since a key can have at most one lock at a time, we don't need to add a timestamp to the key in `CF_LOCK`. Hence the content of each CF:
```

* `CF_DEFAULT`: `(key, start_ts)` -> `value`
* `CF_LOCK`: `key` -> `lock_info`
* `CF_WRITE`: `(key, commit_ts)` -> `write_info`

> Our technical to compose user keys and timestamps together is:

Reviewer 4: Our approach to compound user keys and timestamps together is:

1. Encode the user key to [memcomparable](https://github.com/facebook/mysql-5.6/wiki/MyRocks-record-format#memcomparable-format)

> 2. Bitwise invert the timestamp (which is an unsigned int64) and encode it to big-endian bytes

Reviewer 4: ```suggestion
2. Bitwise invert the timestamp (an unsigned Int64) and encode it into big-endian bytes.
```

Author: I think the term "int64" doesn't need to be capitalized

Reviewer 4: Searched on google and get both instances. So I guess it is fine to not capitalize it.

Reviewer 4: Searched on google and got both instances of capitalized and non-capitalized. So I guess it's fine to leave as is.

3. Append the encoded timestamp to the encoded key.

> For example, the key `"key1"` and the timestamp `3` will be encoded as `"key1\x00\x00\x00\x00\xfb\xff\xff\xff\xff\xff\xff\xff\xfe"`, where the first 9 bytes is the memcomparable-encoded key and the remaining 8 bytes is the inverted timestamp in big-endian. In this way, our different versions of the same key are always adjacent in RocksDB, and for each key, newer versions are always in front of older versions.

Reviewer 4: ```suggestion
For example, key `"key1"` and timestamp `3` will be encoded as `"key1\x00\x00\x00\x00\xfb\xff\xff\xff\xff\xff\xff\xff\xfe"`, where the first 9 bytes is the memcomparable-encoded key and the remaining 8 bytes is the inverted timestamp in big-endian. In this way, different versions of the same key are always adjacent in RocksDB; and for each key, newer versions are always before older ones.
```

Author: "... are always positionally before older ones"?

Reviewer 4: I think the context is enough to indicate "positionally".

Reviewer 4: I think the context is enough to indicate "positionally"

> There are some differences between TiKV and the Percolator's paper. In TiKV, records in `CF_WRITE` has four different types: `Put`, `Delete`, `Rollback` and `Lock`. Only `Put` records need a corresponding value in `CF_DEFAULT`. When rolling back transactions, we don't simply remove the lock but writes a `Rollback` record in `CF_WRITE`. The `Lock` type of write records is different from Percolator's lock, but is produced by queries like `SELECT ... FOR UPDATE` in TiDB. For keys affected by this query, although they are only read, it's part of a write operation. To guarantee to be in snapshot-isolation, we make it acts like a write operation (though it doesn't write anything) to ensure the keys are locked and won't change before committing the transaction.

Reviewer 4: Two confusions to clear:

"The `Lock` type of write records is different from Percolator's lock, **but** is produced by queries like `SELECT ... FOR UPDATE` in TiDB. "  Not sure I see a "but" logic here.

Reviewer 4: "For keys affected by this query, although they are only read, **it's** part of a write operation. "
What does "it" refer to? @MyonKeminta  Please check this.

Author: Is it ok to say "The `Lock` type of write records is not related to Percolator's lock, but..."?
Here I want to say something like "不是...而是..."

Reviewer 4: Yeah this sounds better and clearer in logic.

Reviewer 4: Yeah it sounds more logical to me.

Author: > "For keys affected by this query, although they are only read, **it's** part of a write operation. "
> What does "it" refer to?

Sorry, here I meant "they are", not "it's". It refers to "keys affected by this query".

# Percolator

TiKV supports distributed transactions, which is inspired by Google's [Percolator](https://ai.google/research/pubs/pub36726.pdf). In this section, we will briefly introduce Percolator and how we make use of it in TiKV.

## What is Percolator?

*Percolator* is a system built by Google for incremental processing on a very large data set. Since this is just a brief introduction, you can view the full paper [here](https://ai.google/research/pubs/pub36726#) for more details. If you are already very familiar with it, you can skip this section and go directly to [Percolator in TiKV](#Percolator-in-TiKV)

Percolator is built based on Google's BigTable, a distributed storage system that supports single-row transactions. Percolator implements distributed transactions in ACID snapshot-isolation semantics, which is not supported by BigTable. A column `c` of Percolator is actually divided into the following internal columns of BigTable:

* `c:lock`
* `c:write`
* `c:data`
* `c:notify`
* `c:ack_O`

> Percolator also relies on a service named *timestamp oracle*. The timestamp oracle can produce timestamps in a strictly increasing order. All read and write operations need to apply for timestamps from the timestamp oracle, and a timestamp coming from the timestamp oracle will be regarded as the logical time when the read/write operation happens.

Reviewer: I don't see the paper mentions the concept of `logical time`, we'd better not use this concept?

Author: Emmm... Do you have any better idea about how to describe this? Is it ok to just remove the word `logical`?

| Joe | 7:$9<br/>6:<br/>5:$2 | 7:primary@Bob.bal<br/>6:<br/>5: | 7:<br/>6:data@5<br/>5: |

Then `Commit`:

1. Get the `commit_ts`, in our case, `8`.
2. Commit the primary: Remove the primary lock and write the commit record to the `write` column.

| key | bal:data | bal:lock | bal:write |
|-----|----------|----------|-----------|
| Bob | 8:<br/>7:$3<br/>6:<br>5:$10 | 8:<br/>7:<br/>6:<br/>5: | 8:data@7<br/>7:<br/>6:data@5<br/>5: |
| Joe | 7:$9<br/>6:<br/>5:$2 | 7:primary@Bob.bal<br/>6:<br/>5: | 7:<br/>6:data@5<br/>5: |

3. Commit all secondary locks to complete the writing process.

| key | bal:data | bal:lock | bal:write |
|-----|----------|----------|-----------|
| Bob | 8:<br/>7:$3<br/>6:<br>5:$10 | 8:<br/>7:<br/>6:<br/>5: | 8:data@7<br/>7:<br/>6:data@5<br/>5: |
| Joe | 8:<br/>7:$9<br/>6:<br/>5:$2 | 8:<br/>7:<br/>6:<br/>5: | 8:data@7<br/>7:<br/>6:data@5<br/>5: |

### Reading

Reading from Percolator also requires a timestamp. The procedure to perform a read operation is as follows:

1. Get a timestamp, `ts`.
2. Check if the row we are going to read is locked with a timestamp in the range `[0, ts]`;

>     * If there is a timestamp in range `[0, ts]`, it means the row is locked by an earlier-started transaction. Then we are not sure whether that transaction will be committed before or after `ts`.

Reviewer: Do we need to mention how to deal with this situation? e.g. back off or clean up the lock?

To roll back a row, just simply remove its lock and its corresponding value in `data` column.

### Tolerating crashes

Percolator has the ability to survive crashes without breaking data integrity.

First, let's see what will happen after a crash. A crash may happen during `Prewrite`, `Commit` or between these two phases. We can simply divide these conditions into two types: before committing the primary, or after committing the primary.

So, when a transaction `T1` (either reading or writing) finds that a row `R1` has a lock which belongs to an earlier transaction `T0`, `T1` doesn't simply rollback itself immediately. Instead, it checks the state of `T0`'s primary lock.

* If the primary lock has disappeared and there's a record `data @ T0.start_ts` in the `write` column, it means that `T0` has been successfully committed. Then row `R1`'s stale lock can also be committed. Usually we call this `rolling forward`. After this, the new transaction `T1` resumes.
* If the primary lock has disappeared with nothing left, it means the transaction has been rolled back. Then row `R1`'s stale lock should also be rolled back. After this, `T1` resumes.

> * If the primary lock exists but it's too old (out of TTL set to the lock), it indicates that the transaction has crashed before being committed or rolled back. Roll back `T1` and it will resume.

Reviewer: How about using`wall time` to keep consistent with the paper? BTW, do we need to mention `Chubby` here?

Percolator is built based on Google's BigTable, a distributed storage system that supports single-row transactions. Percolator implements distributed transactions in ACID snapshot-isolation semantics, which is not supported by BigTable. A column `c` of Percolator is actually divided into the following internal columns of BigTable:

* `c:lock`
* `c:write`
* `c:data`
* `c:notify`
* `c:ack_O`

Percolator also relies on a service named *timestamp oracle*. The timestamp oracle can produce timestamps in a strictly increasing order. All read and write operations need to apply for timestamps from the timestamp oracle, and a timestamp coming from the timestamp oracle will be regarded as the logical time when the read/write operation happens.

Percolator is a multi-version storage, and a data item's version is represented by the timestamp when the transaction was committed.

> For example, in such a state:

Reviewer 4: ```suggestion
For example, 
```

| key | v:data | v:lock | v"write |
|-----|--------|--------|---------|
|k1   |14:"value2"<br/>12:<br/>10:"value1"|14:primary<br/>12:<br/>10:|14:<br/>12:data@10<br/>10:|

The state in the table means that for key `k1`, value `"value1"` was committed at timestamp `12`. Then there is an uncommitted version whose value is `"value2"`, and it's uncommitted because there's a lock. You will understand why it is like this after understanding how transactions work.

> The data of a specific row and a specific Percolator column is named a *cell*. The above example shows different versions of data for a single cell.

Reviewer 4: ```suggestion
The data of a specific row and the corresponding column in Percolator is named a *cell*. The above example shows different versions of data for a single cell.
```

Author: The column is not the corresponding thing to the row. A *cell* is of a row and is also of a Percolator's column. Is there any better expression?

The remaining columns, `c:notify` and `c:ack_O`, are used for Percolator's incremental processing. After a modification, `c:notify` column is used to mark the modified cell to be dirty. Users can add some *observers* to Percolator which can do user-specified operations when they find data of their observed columns has changed. To find whether data is changed, the observers continuously scan the `notify` columns to find dirty cells. `c:ack_O` is the "acknowledgment" column of observer `O`, which is used to prevent a row from being incorrectly notified twice. It saves the timestamp of the observer's last execution.

### Writing

Percolator's transactions are committed by a 2-phase commit (2PC) algorithm. Its two phases are `Prewrite` and `Commit`.

In `Prewrite` phase:

> 1. Get a timestamp from the timestamp oracle, and we call the timestamp the transaction's `start_ts`.

Reviewer 4: ```suggestion
1. Get a timestamp from the timestamp oracle, and we call the timestamp `start_ts` of the transaction.
```

2. For each row involved in the transaction, put a lock in the `lock` column. At the same time, write the value to the `data` column with the timestamp `start_ts`. One of these locks will be chosen as the *primary* lock, while others are *secondary* locks. Each lock contains the transaction's `start_ts`. Each secondary lock, in addition, contains the location of the primary lock.
    * If there's already a lock or newer version than `start_ts`, the current transaction will be rolled back.

And then, in the`Commit` phase:
1. Get another timestamp, namely `commit_ts`.

> 2. Remove the primary lock, and at the same time write a record to the `write` column with timestamp `commit_ts`, whose value records the transaction's `start_ts`. If the primary lock is missing, the commit should fail.

Reviewer 4: ```suggestion
2. Remove the primary lock, and at the same time write a record to the `write` column with timestamp `commit_ts`, whose value records the transaction's `start_ts`. If the primary lock is missing, the commit fails.
```

> 3. Then for all secondary locks, repeat the process above.

Reviewer 4: ```suggestion
3. Repeat the process above for all secondary locks.
```

> Once the step 2 (committing the primary) is done, we say the whole transaction is done. It doesn't matter if the process of committing the secondaries unexpectedly exited.

Reviewer 4: ```suggestion
Once step 2 (committing the primary) is done, the whole transaction is done. It doesn't matter if the process of committing the secondary locks unexpectedly failed.
```

Author: If we use the word "fail" here, I think the word "unexpectedly" can also be omitted.

Reviewer 4: Correct. Fail is always unexpected :)

| Bob | 7:$3<br/>6:<br>5:$10 | 7:primary<br/>6:<br/>5:         | 7:<br/>6:data@5<br/>5: |
| Joe | 7:$9<br/>6:<br/>5:$2 | 7:primary@Bob.bal<br/>6:<br/>5: | 7:<br/>6:data@5<br/>5: |

Then `Commit`:

1. Get the `commit_ts`. We got `8`.
2. Commit the primary: Remove the primary lock and put the commit record to `write` column. We get this:

| key | bal:data | bal:lock | bal:write |
|-----|----------|----------|-----------|
| Bob | 8:<br/>7:$3<br/>6:<br>5:$10 | 8:<br/>7:<br/>6:<br/>5: | 8:data@7<br/>7:<br/>6:data@5<br/>5: |
| Joe | 7:$9<br/>6:<br/>5:$2 | 7:primary@Bob.bal<br/>6:<br/>5: | 7:<br/>6:data@5<br/>5: |

3. Commit all secondaries. Finally, we get this:

| key | bal:data | bal:lock | bal:write |
|-----|----------|----------|-----------|
| Bob | 8:<br/>7:$3<br/>6:<br>5:$10 | 8:<br/>7:<br/>6:<br/>5: | 8:data@7<br/>7:<br/>6:data@5<br/>5: |
| Joe | 8:<br/>7:$9<br/>6:<br/>5:$2 | 8:<br/>7:<br/>6:<br/>5: | 8:data@7<br/>7:<br/>6:data@5<br/>5: |

### Reading

To read from Percolator, a timestamp is required too. The procedure to perform read operation is as following:

1. Get a timestamp, `ts`.
2. Check if the row we are going to read is locked with a timestamp in the range `[0, ts]`;

>     * If so, it means the row is locked by an earlier-started transaction. Then we are not sure whether that transaction will be committed before `ts` or after `ts`. Other wise, we can continue reading.

Reviewer 4: ```suggestion
    * If there is a timestamp in range `[0, ts]`, it means the row is locked by an earlier-started transaction. Then we are not sure whether that transaction will be committed before or after `ts`. 
    * If there is not such a timestamp, the read can continue.
```

> 3. Get the latest record in the row's `write` column whose `commit_ts` is in range `[0, ts]`. The record contains the `start_ts` of the transaction where it was committed.

Reviewer 4: ```suggestion
3. Get the latest record in the row's `write` column whose `commit_ts` is in range `[0, ts]`. The record contains the `start_ts` of the transaction when it was committed.
```

> 4. Get the row's value in `data` column, whose timestamp is exactly `start_ts`. Then the value is what we want.

Reviewer 4: ```suggestion
4. Get the row's value in the `data` column whose timestamp is exactly `start_ts`. Then the value is what we want.
```

For example, consider this table again:

| key | bal:data | bal:lock | bal:write |
|-----|----------|----------|-----------|
| Bob | 8:<br/>7:$3<br/>6:<br>5:$10 | 8:<br/>7:<br/>6:<br/>5: |8:data@7<br/>7:<br/>6:data@5<br/>5: |
| Joe | 7:$9<br/>6:<br/>5:$2 | 7:primary@Bob.bal<br/>6:<br/>5: | 7:<br/>6:data@5<br/>5: |

Let's read Bob's balance.

> 1. Get a timestamp. Assume we got `9`.

Reviewer 4: ```suggestion
1. Get a timestamp. Assume it's `9`.
```

> 2. Check lock. The row of Bob is not locked, so we continue.

Reviewer 4: ```suggestion
2. Check the lock of the row. The row of Bob is not locked, so we continue.
```

> 3. Get the latest record in `write` column that committed before `9`. We got the record with `commit_ts` equals to `8`, and it says the `start_ts` is `7`, which means, it's corresponding data is at timestamp `7` in `data` column.

Reviewer 4: ```suggestion
3. Get the latest record in the `write` column committed before `9`. We get a record with `commit_ts` equals to `8`, and`start_ts` `7`, which means, its corresponding data is at timestamp `7` in the `data` column.
```

> 4. Get the value in `data` column whose timestamp is exactly `7`. We got `$3`.

Reviewer 4: ```suggestion
4. Get the value in the `data` column with timestamp `7`. `$3` is the result to the read.
```

> This algorithm provides us both lock-free reading and the ability to perform historical read. In the above example, if we specify that we want to read at time point `7`, then we will see the write record at timestamp `6`, giving us the result `$10` at timestamp `5`.

Reviewer 4: ```suggestion
This algorithm provides us with the abilities of both lock-free read and historical read. In the above example, if we specify that we want to read at time point `7`, then we will see the write record at timestamp `6`, giving us the result `$10` at timestamp `5`.
```

### Handling Conflicts

Conflicts are identified by checking the `lock` column. A row can have many versions of data, but it can have at most one lock at any time.

> When we are performing a write operation, we try to lock every affected row in the `Prewrite` phase. If we failed to lock some of these rows, the whole transaction will be rolled back. As an optimistic lock algorithm, sometimes it causes a shortage of Percolator: there may be performance regression in scenarios where conflicts occur frequently.

Reviewer 4: ```suggestion
When we are performing a write operation, we try to lock every affected row in the `Prewrite` phase. If we failed to lock some of these rows, the whole transaction will be rolled back. As an optimistic lock algorithm, Percolator's transactional write algorithm may cause performance regressions in scenarios where conflicts occur frequently.
```

Author: Seems "cause" is not very accurate here. How about:
> ... As an optimistic lock algorithm, sometimes Percolator's transactional write algorithm may meet performance regressions in scenarios where conflicts occur frequently.

Reviewer 4: What about 
> Using an optimistic lock algorithm, sometimes Percolator's transactional write may encounter performance regressions in scenarios where conflicts occur frequently.

This way we don't need to repeat "algorithm" twice.

> When we are performing a write operation, we try to lock every affected row in the `Prewrite` phase. If we failed to lock some of these rows, the whole transaction will be rolled back. As an optimistic lock algorithm, sometimes it causes a shortage of Percolator: there may be performance regression in a scenario that conflicts occur frequently.

Reviewer 4: ```suggestion
When we are performing a write operation, we try to lock every affected row in the `Prewrite` phase. If we failed to lock some of these rows, the whole transaction will be rolled back. As an optimistic lock algorithm, sometimes it causes a shortage of Percolator: there may be performance regression in scenarios where conflicts occur frequently.
```

Reviewer 4: "As an optimistic lock algorithm, sometimes it causes a shortage of Percolator"
What does "it" refer to here specifically?  And what is its relationship with "an optimistic lock algorithm"？

Author: "it" is Percolator's transactional write algorithm. Percolator's transactional write algorithm is an "optimistic locking algorithm".

To roll back a row, just simply remove its lock and its corresponding value in `data` column.

### Tolerating crashes

> Percolator has the ability to survive crashes without breaking the correctness of data.

Reviewer 4: ```suggestion
Percolator has the ability to survive crashes without compromising data integrity.
```

Author: Can I use "breaking" instead of "compromising"?

Reviewer 4: yeah it is fine

Reviewer 4: Sure. It's just a personal preference. Both are correct.

> First, let's see what will happen after a crash. A crash may happen during `Prewrite`, during `Commit` or between these two. But anyway, we can simply divide these conditions into two types: before committing the primary, or after committing the primary.

Reviewer 4: ```suggestion
First, let's see what will happen after a crash. A crash may happen during `Prewrite`, `Commit` or between these two phases. We can simply divide these conditions into two types: before committing the primary, or after committing the primary.
```

> So, when a transaction `T1` (either reading or writing) finds that a row `R1` has a lock which belongs to an earlier transaction `T0`, it, in fact, doesn't simply rollback itself immediately. However, it checks the state of the lock's primary.

Reviewer 4: ```suggestion
So, when a transaction `T1` (either reading or writing) finds that a row `R1` has a lock which belongs to an earlier transaction `T0`, T1 doesn't simply rollback itself immediately. Instead, it checks the state of the lock's primary.
```

Reviewer 4: the state of "the lock's primary" or "the primary lock"?

Author: Sorry, T0's primary lock.

> * If the primary lock disappeared but there's a record `data @ T0.start_ts` in `write` column, it means that `T0` has been successfully committed. Then the row `R1`'s stale lock can also be committed. Usually we call this `rolling forward`. After that, the new transaction `T1` can continue.

Reviewer 4: ```suggestion
* If the primary lock has disappeared and there's a record `data @ T0.start_ts` in the `write` column, it means that `T0` has been successfully committed. Then row `R1`'s stale lock can also be committed. Usually we call this `rolling forward`. After this, the new transaction `T1` resumes.
```

> * If the primary lock disappeared with nothing left, it means the transaction has been rolled back. Then the row `R1`'s stale lock should also be rolled back. Then `T1` can continue.

Reviewer 4: ```suggestion
* If the primary lock has disappeared with nothing left, it means the transaction has been rolled back. Then row `R1`'s stale lock should also be rolled back. After this, `T1` resumes.
```

> * If the primary lock exists but it's too old (we can set a TTL to the lock), we can believe that the transaction crashed before committing or rolling back. We can rollback it and continue `T1`.

Reviewer 4: ```suggestion
* If the primary lock exists but it's too old (out of TTL set to the lock), it indicates that the transaction has crashed before being committed or rolled back.  Roll back `T1` and it will resume.
```

> * Otherwise, we consider transaction `T0` is still running. Then `T1` can rollback itself, or another choice, try to wait for a while to see whether `T0` will be committed before `T1.start_ts`.

Reviewer 4: ```suggestion
* Otherwise, we consider transaction `T0` to be still running. `T1` can rollback itself, or try to wait for a while to see whether `T0` will be committed before `T1.start_ts`.
```
